File: frontend/src/api/sales.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { apiClient } from './client'

export interface Product {
  id: number
  code: string
  name: string
  unit_price: number
  vat_rate: number
  unit: string | null
  is_active: boolean
}

export interface ProductInput {
  code: string
  name: string
  unit_price: number
  vat_rate: number
  unit?: string | null
}

export interface OrderItem {
  id: number
  product_id: number
  product_code: string | null
  product_name: string | null
  quantity: number
  unit_price: number
  line_total: number
}

export interface OrderItemInput {
  product_id: number
  quantity: number
  unit_price?: number | null
}

export interface SalesOrder {
  id: number
  order_no: string
  customer_id: number
  customer_name: string | null
  order_date: string
  total_amount: number
  status: string
  created_by_name: string | null
  created_at: string
}

export interface SalesOrderDetail extends SalesOrder {
  items: OrderItem[]
}

export interface OrderInput {
  customer_id: number
  status: string
  items: OrderItemInput[]
}

export function useProducts() {
  return useQuery({
    queryKey: ['products'],
    queryFn: async () => {
      const { data } = await apiClient.get<Product[]>('/products')
      return data
    },
  })
}

export function useCreateProduct() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: async (payload: ProductInput) => {
      const { data } = await apiClient.post<Product>('/products', payload)
      return data
    },
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ['products'] }),
  })
}

export function useUpdateProduct() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: async ({
      productId,
      payload,
    }: {
      productId: number
      payload: Partial<ProductInput> & { is_active?: boolean }
    }) => {
      const { data } = await apiClient.patch<Product>(`/products/${productId}`, payload)
      return data
    },
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ['products'] }),
  })
}

export function useDeactivateProduct() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: async (productId: number) => {
      await apiClient.delete(`/products/${productId}`)
    },
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ['products'] }),
  })
}

export function useOrders(params?: { customer_id?: number; status?: string }) {
  return useQuery({
    queryKey: ['orders', params],
    queryFn: async () => {
      const { data } = await apiClient.get<SalesOrder[]>('/orders', { params })
      return data
    },
  })
}

export function useOrderDetail(orderId: number | null) {
  return useQuery({
    queryKey: ['orders', orderId, 'detail'],
    queryFn: async () => {
      const { data } = await apiClient.get<SalesOrderDetail>(`/orders/${orderId}`)
      return data
    },
    enabled: orderId != null,
  })
}

export function useCreateOrder() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: async (payload: OrderInput) => {
      const { data } = await apiClient.post<SalesOrderDetail>('/orders', payload)
      return data
    },
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ['orders'] }),
  })
}

export function useUpdateOrderStatus() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: async ({ orderId, status }: { orderId: number; status: string }) => {
      const { data } = await apiClient.patch<SalesOrder>(`/orders/${orderId}`, { status })
      return data
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['orders'] })
      queryClient.invalidateQueries({ queryKey: ['orders', undefined, 'detail'] })
    },
  })
}

File: backend/app/schemas/sales.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict


class ProductBase(BaseModel):
    code: str
    name: str
    unit_price: float
    vat_rate: float = 20
    unit: str | None = None


class ProductCreate(ProductBase):
    pass


class ProductUpdate(BaseModel):
    code: str | None = None
    name: str | None = None
    unit_price: float | None = None
    vat_rate: float | None = None
    unit: str | None = None
    is_active: bool | None = None


class ProductOut(ProductBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    is_active: bool = True


class SalesOrderItemIn(BaseModel):
    product_id: int
    quantity: float = 1
    unit_price: float | None = None


class SalesOrderItemOut(BaseModel):
    id: int
    product_id: int
    product_code: str | None = None
    product_name: str | None = None
    quantity: float
    unit_price: float
    line_total: float


class SalesOrderCreate(BaseModel):
    customer_id: int
    order_date: date | None = None
    status: str = "draft"
    items: list[SalesOrderItemIn] = []


class SalesOrderUpdate(BaseModel):
    status: str | None = None


class SalesOrderOut(BaseModel):
    id: int
    order_no: str
    customer_id: int
    customer_name: str | None = None
    order_date: date
    total_amount: float
    status: str
    created_by_name: str | None = None
    created_at: datetime


class SalesOrderDetailOut(SalesOrderOut):
    items: list[SalesOrderItemOut] = []
